zyte_common_items/util.py

```diff
@@ -69,3 +70,46 @@ def url_to_str(url: Union[str, _Url]) -> str:
 
 def format_datetime(dt):
     return f"{dt.isoformat(timespec='seconds')}Z"
+
+
+def convert_to_class(value: Any, new_cls: type) -> Any:
+    if type(value) == new_cls:
+        return value
+    input_attributes = {attribute.name for attribute in attrs.fields(value.__class__)}
+    output_attributes = {attribute.name for attribute in attrs.fields(new_cls)}
+    shared_attributes = input_attributes & output_attributes
+    new_value = new_cls(
+        **{attribute: getattr(value, attribute) for attribute in shared_attributes}
+    )
+    removed_nonempty_attributes = {
+        attribute
+        for attribute in (input_attributes - output_attributes)
+        if getattr(value, attribute) is not None
```

Comment:
I think what "empty" means can vary. For example, removing probability=1 is fine if probability is not needed.

Reply:
Do you want to explicitly check and ignore the probability=1 case here? Or try to find if the field value is the default one?

Reply:
I'm not sure what's the best implementation. I think we should ensure that if user passes Metadata instance as a value of some metadata attribute, then 

1) if probability is allowed for this attribute, then it's set to 1, unless user set it to a different value explicitly
2) if probability is not allowed, and user doesn't set probability explicitly when passing Metadata instance, there shouldn't be a warning
2) if probability is not allowed, but user sets it explicitly to a value which is not 1 (or 0? no sure about this), then a warning might be logged. 

I'm not 100% sure about (3), but I think we should ensure (1) and (2) work. My worry here was that if we add a default probability=1 to Metadata, then the current code will start showing warnings.

Reply:
I don't think we know here if the user passed the probability value explicitly? So we can only compare it with the default, and log the warning only if the value is not the default one? That's 2+3

Reply:
What I mean is this:

```py
metadata=Metadata()
```

So, (1) means that metadata.probability should be 1 here; if probability=1 is used as a default value in Metadata.probability, that's achieved.

(2) means that if metadata attribute doesn't have probability, there shouldn't be a warning for in the example.

(3) implies a different code example:

```py
metadata=Metadata(probability=0.5)
```

^^ In this case it does make sense to have a warning. To address the warning, user should remove probability argument, converting it to the 1st example.

Then, there is a question on what to do in this case, when the probability is not supported:

```py
metadata=Metadata(probability=1)
```

I think it's fine both to ignore it (don't raise a warning), and to raise a warning.

Reply:
> I think it was the case before probability=1 was added as a default to Metadata class, but not the case now - without further work by addressing (1) we can break (2) .

Yes, assuming `MyItem.metadata` doesn't have `probability` (e.g. when `MyItem` is `ProductList`) we get a warning. How should we fix it? Add a check that if the removed attribute is `probability` and its value is `1.0` then don't count it for the warning?

Reply:
(Which was even easier than I thought, pushed)